src/amber/cli/helpers/process_runner.cr

```diff
@@ -1,114 +1,233 @@
       if file_counter > 0
-        log "Watching #{file_counter} files (server reload)..."
-        start_app
+        debug "Watching #{file_counter} #{key} files"
+        kill_processes(key)
+        commands.each do |command|
+          if key == "server" && command == commands.first?
+            run_build_command(command, commands)
+          else
+            run_command(command, key)
+          end
+        end
       end
     end
 
-    private def stop_all_processes
-      log "Terminating app #{project_name}..."
-      @processes.each do |process|
-        process.kill unless process.terminated?
+    private def check_directories
+      Dir.mkdir_p("bin")
+      if !Dir.exists?("lib")
+        error "You need to install dependencies first, execute `shards install`"
+        exit 1
       end
-      processes.clear
     end
 
-    private def create_all_processes
-      process = create_watch_process
-      @processes << process if process.is_a? Process
-      unless @npm_process
-        create_npm_process
-        @npm_process = true
+    private def run_build_command(command, commands)
+      check_directories
+      next_server_commands_range = (1...commands.size)
+      info "Building project #{Helpers.settings.name.colorize(:light_cyan)}"
+      spawn do
+        error_io = IO::Memory.new
+        process = Helpers.run(command, error: error_io)
+        PROCESSES << {process, "server"}
+        loop do
+          if process.terminated?
+            handle_terminaded_process(process, error_io)
+            break
+          end
+          sleep 1
+        end
       end
     end
 
-    private def build_app_process
-      log "Building project #{project_name}..."
-      Amber::CLI::Helpers.run(@build_command)
+    private def handle_terminaded_process(process, error_io)
+      exit_status = process.wait.exit_status
+      if error_io.empty?
+        if exit_status.zero?
+          if @watch_running
+            kill_processes("server")
+          else
+            notify_next_processes
+          end
+          next_server_commands_range.each { @wait_build.send true }
+        else
+          next_server_commands_range.each { @wait_build.send false }
+        end
+      else
+        handle_error(error_io.to_s)
+        next_server_commands_range.each { @wait_build.send false }
+      end
     end
 
-    private def create_watch_process
-      log "Starting #{project_name}..."
-      Amber::CLI::Helpers.run(@run_command, wait: false, shell: false)
+    private def notify_next_processes
+      notify_counter = @notify_counter_channel.receive
+      notify_counter.times { @notify_channel.send nil }
+      @notify_counter_channel.send 0
     end
 
-    private def create_npm_process
-      node_log "Installing dependencies..."
-      Amber::CLI::Helpers.run("npm install --loglevel=error && npm run watch", wait: false)
-      node_log "Watching public directory"
+    private def run_command(command, key)
+      if key == "server"
```

Comment:
Why is "server" special?

Reply:
Yeah, I did such a complex stuff here, I'm already fixing this :+1: